src/lib/backend.c

```diff
@@ -190,11 +184,26 @@ CK_RV backend_get_tokens(token **tok, size_t *len) {
                 rv = CKR_OK;
             }
         }
+        LOGV("FAPI + Esysdb returned %zi token", *len);
     }
 
-    if (esysdb_init) {
-        (*tok)[*len] = tmp;
-        *len += 1;
+    if (*len >= MAX_TOKEN_CNT) {
+        LOGW("Too many tokens, must have less than %d to show empty tokens", MAX_TOKEN_CNT);
+        return CKR_OK;
+    }
+
+    token *t = &(*tok)[*len];
+    t->id = *len + 1;
+    for (size_t i = 0; i < *len; i++) {
+        if (((*tok)[i]).id > t->id) {
+            t->id = ((*tok)[i]).id + 1;
+        }
+    }
```

Comment:
So what's the purpose of this? In case FAPI and ESYSDB have token id as the same (replace the 0x80 + removed below)?

It probably is even better to update `token *slot_get_token(CK_SLOT_ID slot_id)` where we don't use token->id to map slot to token, we can just use the index location in the list. Then id is never really used for anything but perhaps debug statements and sql queries.

Reply:
The problem was, that if the token id comes from esysdb and the fapi-id is just esysdb + 0x80 then we will have a collision of ids within fapi-backend if we create a second token on fapi, because esysdb does not know about fapi's used ids.

Reply:
make it so id doesn't matter to pkcs11 internals. Then it's only consumed by backends, which is already split on the backend selector enum. Then we don't care about its value.

Reply:
I pretty much do so right now by looking for an id not consumed by either of the backend.
I cannot really remove the tokenid from the esysdb backend because it's used a primary key of the table.
Or should I decouple this ?

Reply:
> I pretty much do so right now by looking for an id not consumed by either of the backend.
> I cannot really remove the tokenid from the esysdb backend because it's used a primary key of the table.
> Or should I decouple this ?

The latter. We should be able to keep the backend code in place as is. But this id is also used for slot to token mapping. We can decouple it there, and then collisions really don't matter. The main logic is in slot.c:
```C
token *slot_get_token(CK_SLOT_ID slot_id) {

    slot_lock();

    size_t i;
    for (i=0; i < global.token_cnt; i++) {
        token *t = &global.token[i];
        if (slot_id == t->id) {
            slot_unlock();
            return t;
        }
    }

    slot_unlock();
    return NULL;
}
```
If we just map slot id to index in token list, we should be good. It would also fix some weirdness with slots possible having gaps, if token 1, 4, 7 were to exist.

The only time collisions might matter is debug print statements, we would need to print both id and back-end selector to distinguish them in the high level pkcs11 code. The backend print statements you can increase the log level and get filename and imply backend from that.

Reply:
Ok, so maybe in the mean time we could merge this PR so we have support for multiple tokens via FAPI backend (which is a current bug) and refactoring the id-code comes in the next step (after I debugged the tabrmd-problem with 0x902s) ?

Reply:
Theirs a bug where esysdb hands out tokid 255 and you make it 256, which will overflow a byte. IIRC correctly, I map the the session handles to tokens using the top byte as an identifier to the token. So I am leaning towards no, were swapping one bug for another.

Reply:
Done